cadc-pkg-server/src/main/java/org/opencadc/pkg/server/TarWriter.java

```diff
@@ -0,0 +1,257 @@
+import ca.nrc.cadc.net.HttpDownload;
+import java.io.ByteArrayOutputStream;
+import java.io.FilterOutputStream;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.io.PrintWriter;
+import java.net.URI;
+import java.net.URL;
+import java.util.ArrayList;
+import java.util.Date;
+import java.util.List;
+import java.util.Map;
+import java.util.TreeMap;
+import org.apache.commons.compress.archivers.ArchiveEntry;
+import org.apache.commons.compress.archivers.tar.TarArchiveEntry;
+import org.apache.commons.compress.archivers.tar.TarArchiveOutputStream;
+import org.apache.log4j.Logger;
+
+/**
+ *
+ * @author pdowler
+ */
+public class TarWriter
+{
+    private static final Logger log = Logger.getLogger(TarWriter.class);
+    
+    public static final String CONTENT_TYPE = "application/x-tar";
+    
+    private final TarArchiveOutputStream tout;
+    private final Map<String,List<TarContent>> map = new TreeMap<String,List<TarContent>>();
+    
+    public TarWriter(OutputStream ostream) 
+    { 
+        this.tout = new TarArchiveOutputStream(ostream);
+        tout.setLongFileMode(TarArchiveOutputStream.LONGFILE_POSIX);
+    }
+    
+    private class TarContent
+    {
+        URL url;
+        String filename;
+        String contentMD5;
+        String emsg;
+    }
+    
+    public void close()
+        throws IOException
+    {
+        StringBuilder sb = new StringBuilder();
+        for (Map.Entry<String,List<TarContent>> me : map.entrySet())
+        {
+            for (TarContent tc : me.getValue())
+            {
+                if (tc.emsg != null)
+                    sb.append("ERROR ").append(tc.emsg);
+               else
+                    sb.append("OK ").append(tc.filename).append(" ").append(tc.contentMD5).append(" ").append(tc.url);
+                sb.append("\n");
+            }
+            boolean openEntry = false;
+            try
+            {
+                String filename = me.getKey() + "/README";
+                tout.putArchiveEntry(new DynamicTarEntry(filename, sb.length(), new Date()));
+                openEntry = true;
+                PrintWriter pw = new PrintWriter(new TarStreamWrapper(tout));
+                pw.print(sb.toString());
+                pw.flush();
+                pw.close(); // safe
+            }
+            finally
+            {
+                if (openEntry)
+                    tout.closeArchiveEntry();
+            }
+        }
+        
+        tout.finish();
+        tout.close();
+    }
+    
+    public void addMessage(String path, String msg)
+    {
+        TarContent tc = new TarContent();
+        tc.emsg = msg;
+        addItem(path, tc);
+    }
+    
+    private void addItem(String path, TarContent tc)
+    {
+        List<TarContent> t = map.get(path);
+        if (t == null)
+        {
+            t = new ArrayList<TarContent>();
+            map.put(path, t);
+        }
+        t.add(tc);
+    }
+    
+    public void write(String path, URL url)
```

Comment:
Should change this method signature to `write(PackageItem)` to show how it's connected to the other classes. It's true that this class would be slightly more generic and still usable without doing that -- the caller could just do
```
write(item.getRelativePath(), item.getURL());
```
However, we found out in cadc-soda-server that when we did something like that we couldn't add extra info without changing the arg list of the method, so it was easier to evolve if we passed a struct and then had the struct evolve.

Reply:
done